### PR TITLE
Update plugin POM

### PR DESCRIPTION
**Pull request by basil**
Supercedes #7 and #9.

- Update Jenkins Core requirement to 2.60.3 and Java 8
- Update Plugin POM to latest version
- Fix Servlet API 3.x binary compatibility issue
- Build on both Java 8 and 11
- Incrementalify the plugin

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@jglick You are the maintainer according to https://github.com/jenkins-infra/repository-permissions-updater/blob/master/permissions/plugin-git-server.yml . Would you be available to cut the release or to grant me permissions?

File: Jenkinsfile [hunk 1]
```diff
@@ -1,4 +1 @@
-#!/usr/bin/env groovy
-
-/* `buildPlugin` step provided by: https://github.com/jenkins-infra/pipeline-library */
-buildPlugin()
+buildPlugin(configurations: buildPlugin.recommendedConfigurations())
```
Comment (oleg-nenashev) [inline Jenkinsfile]: Note that it is not really working in the PR. @basil has no write permissions to the repo

Comment (batmat) [reply Jenkinsfile]: Yep, so as-is this is risky and could break master. I'm otp so cannot replay it myself directly.

Comment (oleg-nenashev) [reply Jenkinsfile]: In the worst case I will do the master surgery. There is no pressing need in the release, so I am not that concerned about breaking master

**Comment by jglick** [comment]
Can check it on Tuesday.

File: Jenkinsfile (hunk 1 above)
Comment (jglick) [reply Jenkinsfile]: Fails in #11 due to `RequireUpperBoundDeps` on `instance-identity`.

**Comment by basil** [comment]
Thanks for running this through CI, @jglick. I reproduced the error locally by running `mvn clean package -DskipTests -Djenkins.version=2.164.1`. My analysis of the situation is as follows. I'm somewhat new to Jenkins modules, so feel free to let me know if I'm misunderstanding something.

`git-server-plugin` imports classes from `org.jenkinsci.main.modules.sshd`. These dependencies are a bit awkward in that they are "[bundled alongside core and cannot be managed by users in the plugin manager](https://jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/)". The documentation goes on to say:

> For Maven dependency management purposes, plugins can declare provided-scope dependencies on these modules if they wish to use their APIs, taking care to select the same version as is bundled in the baseline version of core.

It seems unfortunate that this is necessary. This really reflects a deeper problem. [JENKINS-55582](https://issues.jenkins-ci.org/browse/JENKINS-55582) might possible help here. But in the meantime, what is to be done for this PR? This patch does get Enforcer to pass on both 2.60.3 and 2.164.1:

```
diff --git a/pom.xml b/pom.xml
index c6a0752..0f7082f 100644
--- a/pom.xml
+++ b/pom.xml
@@ -43,7 +43,7 @@
     <dependency>
       <groupId>org.jenkins-ci.modules</groupId>
       <artifactId>instance-identity</artifactId>
-      <version>2.1</version>
+      <version>2.2</version>
       <scope>provided</scope>
     </dependency>
     <dependency>
@@ -55,7 +55,7 @@
     <dependency>
       <groupId>org.jenkins-ci.modules</groupId>
       <artifactId>sshd</artifactId>
-      <version>1.11</version>
+      <version>2.6</version>
       <scope>provided</scope>
     </dependency>
```

However, I don't feel this code is correct because `instance-identity` 2.2 and `sshd` 2.6 aren't _actually_ provided in 2.60.3, which is the official baseline of this plugin. Instead I opted to add an exclusion to the `RequireUpperBoundDeps` configuration. It's unfortunate that this is necessary, but hopefully it can become unnecessary once JENKINS-55582 is implemented. In the meantime, it seems to be the better of two evils, in that it doesn't result in the wrong versions being used on 2.60.3 for the sake of a build warning.

**Comment by jglick** [comment]
> JENKINS-55582 might possibl[y] help here.

Yes I think https://github.com/jenkinsci/jenkins/pull/3988 would solve this sort of issue in the long run.

**Comment by jglick** [comment]
Alternately, just drop support for 2.60.x, which is after all quite old, and use a newer baseline.

**Comment by basil** [comment]
> Alternately, just drop support for 2.60.x, which is after all quite old, and use a newer baseline.

I can do that if desired. Which baseline would you suggest?

**Comment by jglick** [comment]
Now we are getting [some complex errors](https://ci.jenkins.io/job/Plugins/job/git-server-plugin/job/PR-11/3/testReport/junit/org.jvnet.hudson.test.junit/FailedTest/windows_11_2_164_1___Archive__windows_11_2_164_1____org_jvnet_hudson_test_JellyTestSuiteBuilder$JellyTestSuite/).

> Which baseline would you suggest?

Basically as new as you want. I doubt it matters. Beware that the `recommendedConfigurations` does _not_ work when you have a new LTS baseline—it tries to test against something _older_. See https://github.com/jenkins-infra/pipeline-library/pull/92 / https://github.com/jenkins-infra/pipeline-library/pull/94 where discussion has stalled.

**Comment by basil** [comment]
> Now we are getting [some complex errors](https://ci.jenkins.io/job/Plugins/job/git-server-plugin/job/PR-11/3/testReport/junit/org.jvnet.hudson.test.junit/FailedTest/windows_11_2_164_1___Archive__windows_11_2_164_1____org_jvnet_hudson_test_JellyTestSuiteBuilder$JellyTestSuite/).

That test failure was caused by `java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter`. The JAXB dependency was removed in jenkinsci/instance-identity-module#11, which was released in `instance-identity-2.2`. That change was pulled into Jenkins core in jenkinsci/jenkins#3528, which was released in `jenkins-2.130`. So let's pick 2.138.4 as the baseline. I tested this with Java 11 locally and it seems to work.

**Comment by jglick** [comment]
OK. (We could even go to 2.164.x and not have the hack about `RequireUpperBoundDeps` at all.)

**Comment by basil** [comment]
> (We could even go to 2.164.x and not have the hack about `RequireUpperBoundDeps` at all.)

I suppose, but then we'd just have a similar problem the next time the "recommended" version the `Jenkinsfile` uses is increased. I think the current solution is good enough in practice and the hack can be removed entirely once modules are converted to detached plugins.

With that, the CI build seems to be passing now, so I think we can move forward with this.

**Pull request merged**